graph/src/data/subgraph/mod.rs

```diff
@@ -814,6 +859,27 @@ impl<C: Blockchain> UnvalidatedSubgraphManifest<C> {
             errors.push(SubgraphManifestValidationError::DataSourceBlockHandlerLimitExceeded)
         }
 
+        // For API versions newer than 0.0.5, validate that all mappings uses the same api_version
+        let referential_api_version = Version::new(0, 0, 5);
+        let mappings = self.0.mappings();
+        let unique_api_versions: HashSet<&Version> = mappings
+            .iter()
+            .map(|mapping| &mapping.api_version)
+            .collect();
+        if unique_api_versions.len() > 1
+            && unique_api_versions
+                .iter()
+                .any(|version| *version >= &referential_api_version)
+        {
+            let error_msg = unique_api_versions
+                .iter()
+                .map(ToString::to_string)
+                .join(", ");
+            errors.push(SubgraphManifestValidationError::DifferentApiVersions(
+                error_msg,
+            ));
+        }
+
```

Comment:
I could not reuse the `UnifiedMappingApiVersion` logic in this spot because it would silently resolve into a `None` for invalid cases.  

Also, we're now iterating over `Manifest.mappings(), which yields values both from `.data_sources` and `.templates`.

Reply:
>I could not reuse the UnifiedMappingApiVersion logic in this spot because it would silently resolve into a None for invalid cases.

It doesn't have to be that way, you can make that return a `Result`. We _really_ should avoid duplicating the logic.

Reply:
@tilacog this logic is still duplicated.